File: app/src/main/java/com/originb/inkwisenote/modules/repositories/SmartNotebookRepository.java
package com.originb.inkwisenote.modules.repositories;

import com.originb.inkwisenote.common.DateTimeUtils;
import com.originb.inkwisenote.common.ListUtils;
import com.originb.inkwisenote.modules.smartnotes.data.*;
import com.originb.inkwisenote.common.Strings;

import java.util.*;
import java.util.stream.Collectors;

public class SmartNotebookRepository {

    private final AtomicNoteEntitiesDao atomicNoteEntitiesDao;
    private final SmartBooksDao smartBooksDao;
    private final SmartBookPagesDao smartBookPagesDao;

    public SmartNotebookRepository() {

        this.atomicNoteEntitiesDao = Repositories.getInstance().getNotesDb().atomicNoteEntitiesDao();
        this.smartBooksDao = Repositories.getInstance().getNotesDb().smartBooksDao();
        this.smartBookPagesDao = Repositories.getInstance().getNotesDb().smartBookPagesDao();
    }

    // Create the data and return the notebook entity
    public Optional<SmartNotebook> initializeNewSmartNotebook(String title,
                                                              String directoryPath,
                                                              NoteType noteType) {

        AtomicNoteEntity atomicNoteEntity = newAtomicNote("", directoryPath, noteType);
        SmartBookEntity smartBookEntity = newSmartBook(title, atomicNoteEntity.getCreatedTimeMillis());
        SmartBookPage smartBookPage = newSmartBookPage(smartBookEntity, atomicNoteEntity, 0);
        SmartNotebook smartNotebook = new SmartNotebook(smartBookEntity, smartBookPage, atomicNoteEntity);
        return Optional.ofNullable(smartNotebook);
    }

    public void deleteSmartNotebook(SmartNotebook smartNotebook) {
        // will pages allow this to be deleted first?
        smartNotebook.getAtomicNotes()
                .stream().map(AtomicNoteEntity::getNoteId)
                .forEach(atomicNoteEntitiesDao::deleteAtomicNote);

        smartBookPagesDao.deleteSmartBookPages(smartNotebook.getSmartBook().getBookId());

        smartBooksDao.deleteSmartBook(smartNotebook.getSmartBook().getBookId());
    }

    public void deleteNoteFromBook(SmartNotebook smartNotebook, AtomicNoteEntity atomicNote) {
        // will pages allow this to be deleted first?
        atomicNoteEntitiesDao.deleteAtomicNote(atomicNote.getNoteId());
        smartBookPagesDao.deleteNotePages(atomicNote.getNoteId());

        getSmartNotebooks(smartNotebook.getSmartBook().getBookId())
                .filter(updatedSmartNotebook -> updatedSmartNotebook.atomicNotes.isEmpty())
                .ifPresent(updatedSmartNotebook -> {
                    smartBookPagesDao.deleteSmartBookPages(smartNotebook.smartBook.getBookId());
                    smartBooksDao.deleteSmartBook(smartNotebook.getSmartBook().getBookId());
                });

    }

    public void updateNotebook(SmartNotebook smartNotebook) {
        long updateTime = System.currentTimeMillis();

        SmartBookEntity smartBookEntity = smartNotebook.getSmartBook();
        smartBookEntity.setLastModifiedTimeMillis(System.currentTimeMillis());
        smartBooksDao.updateSmartBook(smartBookEntity);

        for (AtomicNoteEntity atomicNote : smartNotebook.getAtomicNotes()) {
            atomicNote.setLastModifiedTimeMillis(updateTime);
        }
        atomicNoteEntitiesDao.updateAtomicNote(smartNotebook.getAtomicNotes());

        int updateResult = smartBookPagesDao.updateSmartBookPage(smartNotebook.getSmartBookPages());
    }

    public Optional<SmartNotebook> getSmartNotebookContainingNote(long noteId) {
        List<SmartBookPage> pagesOfNote = smartBookPagesDao.getSmartBookPagesOfNote(noteId);
        if (pagesOfNote == null || pagesOfNote.isEmpty()) return Optional.empty();

        // TODO: only get the first page for now. We will fetch more later
        long bookId = pagesOfNote.stream().findFirst().map(SmartBookPage::getBookId).get();

        return getSmartNotebooks(bookId);
    }

    public List<SmartNotebook> getAllSmartNotebooks() {
        List<SmartBookEntity> smartBooks = smartBooksDao.getAllSmartBooks();
        if (smartBooks == null || smartBooks.isEmpty()) return new ArrayList<>();

        List<SmartNotebook> smartNotebooks = new ArrayList<>();

        for (SmartBookEntity smartBook : smartBooks) {
            List<SmartBookPage> smartBookPages = smartBookPagesDao.getSmartBookPages(smartBook.getBookId());
            if (smartBookPages == null || smartBookPages.isEmpty()) continue;

            Set<Long> noteIds = smartBookPages.stream()
                    .map(SmartBookPage::getNoteId)
                    .collect(Collectors.toSet());

            List<AtomicNoteEntity> atomicNoteEntities = atomicNoteEntitiesDao.getAtomicNotes(noteIds);

            smartNotebooks.add(new SmartNotebook(smartBook, smartBookPages, atomicNoteEntities));
        }

        return smartNotebooks;
    }

    public Set<SmartNotebook> getSmartNotebooksForNoteIds(Set<Long> noteIds) {
        List<SmartBookPage> smartBookPages = smartBookPagesDao.getSmartBookPagesOfNote(noteIds);
        Map<Long, List<SmartBookPage>> bookIdToPageMap = ListUtils.groupBy(smartBookPages, SmartBookPage::getBookId);

        List<SmartBookEntity> smartBooks = smartBooksDao.getSmartBooks(bookIdToPageMap.keySet());
        if (smartBooks == null || smartBooks.isEmpty()) return new HashSet<>();

        List<AtomicNoteEntity> atomicNotes = atomicNoteEntitiesDao.getAtomicNotes(noteIds);
        Map<Long, List<AtomicNoteEntity>> noteIdToNotes = ListUtils.groupBy(atomicNotes, AtomicNoteEntity::getNoteId);

        Set<Long> bookIdsToRemove = new HashSet<>();
        Set<SmartNotebook> smartNotebooks = new HashSet<>();
        for (SmartBookEntity smartBook : smartBooks) {
            if (!bookIdToPageMap.containsKey(smartBook.getBookId())) {
                bookIdsToRemove.add(smartBook.getBookId());
                continue;
            }
            List<SmartBookPage> bookPages = bookIdToPageMap.get(smartBook.getBookId());
            List<AtomicNoteEntity> bookNotes = bookPages.stream().map(SmartBookPage::getNoteId)
                    .map(noteIdToNotes::get).filter(Objects::nonNull)
                    .flatMap(List::stream)
                    .collect(Collectors.toList());

            smartNotebooks.add(new SmartNotebook(smartBook, smartBookPages, bookNotes));
        }

        return smartNotebooks;
    }

    public Optional<SmartNotebook> getSmartNotebooks(long bookId) {
        SmartBookEntity smartBook = smartBooksDao.getSmartbooksWithMatchingTitle(bookId);
        if (smartBook == null) return Optional.empty();

        List<SmartBookPage> smartBookPages = smartBookPagesDao.getSmartBookPages(bookId);
        if (smartBookPages == null || smartBookPages.isEmpty()) return Optional.empty();

        Set<Long> noteIds = smartBookPages.stream()
                .map(SmartBookPage::getNoteId)
                .collect(Collectors.toSet());

        List<AtomicNoteEntity> atomicNoteEntities = atomicNoteEntitiesDao.getAtomicNotes(noteIds);

        SmartNotebook smartNotebook = new SmartNotebook(smartBook, smartBookPages, atomicNoteEntities);

        return Optional.ofNullable(smartNotebook);
    }

    public Set<SmartNotebook> getSmartNotebooks(String title) {
        if (title.length() < 3) return new HashSet<>();

        List<SmartBookEntity> smartBooks = smartBooksDao.getSmartbooksWithMatchingTitle("%" + title + "%");
        if (smartBooks == null || smartBooks.isEmpty()) return new HashSet<>();

        Set<Long> bookIds = smartBooks.stream().map(SmartBookEntity::getBookId).collect(Collectors.toSet());
        List<SmartBookPage> smartBookPages = smartBookPagesDao.getSmartBooksPages(bookIds);
        Map<Long, List<SmartBookPage>> bookIdToPagesMap = ListUtils.groupBy(smartBookPages, SmartBookPage::getBookId);

        Set<Long> noteIds = smartBookPages.stream().map(SmartBookPage::getNoteId).collect(Collectors.toSet());
        List<AtomicNoteEntity> atomicNotes = atomicNoteEntitiesDao.getAtomicNotes(noteIds);
        Map<Long, List<AtomicNoteEntity>> noteIdToNotes = ListUtils.groupBy(atomicNotes, AtomicNoteEntity::getNoteId);

        Set<SmartNotebook> smartNotebooks = new HashSet<>();
        Set<Long> bookIdsOfEmptyBooks = new HashSet<>();
        for (SmartBookEntity smartBook : smartBooks) {
            long bookId = smartBook.getBookId();
            if (!bookIdToPagesMap.containsKey(bookId)) {
                bookIdsOfEmptyBooks.add(bookId);
                continue;
            }
            List<SmartBookPage> bookPages = bookIdToPagesMap.get(bookId);
            List<AtomicNoteEntity> bookNotes = bookPages.stream().map(SmartBookPage::getNoteId)
                    .map(noteIdToNotes::get).filter(Objects::nonNull)
                    .flatMap(List::stream)
                    .collect(Collectors.toList());
            smartNotebooks.add(new SmartNotebook(smartBook, bookPages, bookNotes));
        }

        return smartNotebooks;
    }

    public AtomicNoteEntity newAtomicNote(String filename, String filepath, NoteType noteType) {
        long createdTimeMillis = System.currentTimeMillis();
        AtomicNoteEntity atomicNoteEntity = new AtomicNoteEntity();
        atomicNoteEntity.setCreatedTimeMillis(createdTimeMillis);

        if (Strings.isNullOrWhitespace(filename)) {
            atomicNoteEntity.setFilename(String.valueOf(createdTimeMillis));
        } else {
            atomicNoteEntity.setFilename(filename);
        }
        if (Strings.isNullOrWhitespace(filepath)) {
            atomicNoteEntity.setFilepath("");
        } else {
            atomicNoteEntity.setFilepath(filepath);
        }

        atomicNoteEntity.setNoteType(noteType.toString());

        long noteId = atomicNoteEntitiesDao.insertAtomicNote(atomicNoteEntity);
        atomicNoteEntity.setNoteId(noteId);

        return atomicNoteEntity;
    }

    private SmartBookEntity newSmartBook(String title, long createdDateTimeMs) {
        SmartBookEntity smartBookEntity = new SmartBookEntity();
        if (Strings.isNullOrWhitespace(title)) {
            smartBookEntity.setTitle(DateTimeUtils.msToDateTime(createdDateTimeMs));
        } else {
            smartBookEntity.setTitle(title);
        }
        smartBookEntity.setCreatedTimeMillis(createdDateTimeMs);
        smartBookEntity.setLastModifiedTimeMillis(createdDateTimeMs);

        long bookId = smartBooksDao.insertSmartBook(smartBookEntity);
        smartBookEntity.setBookId(bookId);

        return smartBookEntity;
    }

    public SmartBookPage newSmartBookPage(SmartBookEntity smartBookEntity,
                                          AtomicNoteEntity atomicNoteEntity,
                                          int pageOrder) {

        SmartBookPage smartBookPage = new SmartBookPage(smartBookEntity.getBookId(),
                atomicNoteEntity.getNoteId(),
                pageOrder);
        long id = smartBookPagesDao.insertSmartBookPage(smartBookPage);
        smartBookPage.setId(id);

        return smartBookPage;
    }


}

File: androidApp/src/main/java/com/originb/inkwisenote2/modules/repositories/SmartNotebookRepository.kt
package com.originb.inkwisenote2.modules.repositories

import android.app.Application
import android.content.Context
import com.originb.inkwisenote2.common.ListUtils
import com.originb.inkwisenote2.common.isNullOrWhitespace
import com.originb.inkwisenote2.modules.backgroundjobs.Events.*
import com.originb.inkwisenote2.modules.smartnotes.data.*
import org.greenrobot.eventbus.EventBus
import java.io.File
import java.nio.file.Paths
import java.util.*

class SmartNotebookRepository(
    private val atomicNotesDomain: AtomicNotesDomain,
    private val atomicNoteEntitiesDao: AtomicNoteEntitiesDao,
    private val smartBooksDao: SmartBooksDao,
    private val smartBookPagesDao: SmartBookPagesDao
) {
    // Create the data and return the notebook entity
    fun initializeNewSmartNotebook(
        title: String,
        directoryPath: String,
        noteType: NoteType
    ): SmartNotebook {
        // Create a directory with the notebook title
        val notebookDirectory = Paths.get(directoryPath, title).toString()
        val notebookDir = File(notebookDirectory)
        if (!notebookDir.exists()) {
            notebookDir.mkdirs()
        }

        val atomicNoteEntity = atomicNotesDomain.saveAtomicNote(
            AtomicNotesDomain.Companion.constructAtomicNote(
                "",
                notebookDirectory,
                noteType
            )
        )
        val smartBookEntity = newSmartBook(title, atomicNoteEntity.createdTimeMillis)
        val smartBookPage = newSmartBookPage(smartBookEntity, atomicNoteEntity, 0)
        val smartNotebook = SmartNotebook(smartBookEntity, smartBookPage, atomicNoteEntity)
        return smartNotebook
    }

    fun deleteSmartNotebook(smartNotebook: SmartNotebook) {
        // Delete notebook folder from filesystem if it exists
        try {
            if (smartNotebook.atomicNotes.isNotEmpty()) {
                val firstNote = smartNotebook.atomicNotes[0]
                if (firstNote.filepath != null) {
                    val notebookDir = File(firstNote.filepath)
                    if (notebookDir.exists() && notebookDir.isDirectory()) {
                        // Delete all files in the directory
                        val files = notebookDir.listFiles()
                        if (files != null) {
                            for (file in files) {
                                file.delete()
                            }
                        }
                        // Delete the directory itself
                        notebookDir.delete()
                    }
                }
            }
        } catch (e: Exception) {
            // Log error but continue with database deletion
            System.err.println("Error deleting notebook directory: " + e.message)
        }

        // Delete database records
        smartNotebook.smartBook?.let { smartBookPagesDao.deleteSmartBookPages(it.bookId) }

        for (note in smartNotebook.atomicNotes) {
            atomicNoteEntitiesDao.deleteAtomicNote(note.noteId)
        }

        smartNotebook.smartBook?.let { smartBooksDao.deleteSmartBook(it.bookId) }

        EventBus.getDefault().post(NotebookDeleted(smartNotebook))
    }

    fun deleteNoteFromBook(smartNotebook: SmartNotebook, atomicNote: AtomicNoteEntity) {
        if (smartNotebook.atomicNotes.size <= 1) deleteSmartNotebook(smartNotebook)
        else {
            smartBookPagesDao.deleteNotePages(atomicNote.noteId)
            atomicNoteEntitiesDao.deleteAtomicNote(atomicNote.noteId)

            EventBus.getDefault().post(NoteDeleted(smartNotebook, atomicNote))
        }
    }

    fun updateNotebook(smartNotebook: SmartNotebook, context: Context) {
        val updateTime = System.currentTimeMillis()

        val smartBookEntity = smartNotebook.smartBook ?: return

        if (smartBookEntity.bookId == -1L) return

        smartBookEntity.lastModifiedTimeMillis = System.currentTimeMillis()
        smartBooksDao.updateSmartBook(smartBookEntity)

        for (atomicNote in smartNotebook.atomicNotes) {
            atomicNote.lastModifiedTimeMillis = updateTime
        }
        atomicNotesDomain.updateAtomicNotes(smartNotebook.atomicNotes)

        smartBookPagesDao.updateSmartBookPage(smartNotebook.smartBookPages)
        EventBus.getDefault().post(SmartNotebookSaved(smartNotebook, context))
    }

    fun saveSmartNotebook(smartNotebook: SmartNotebook, context: Application?): SmartNotebook {
        val updateTime = System.currentTimeMillis()
        val smartBookEntity = smartNotebook.smartBook ?: return smartNotebook
        if (smartBookEntity.bookId > -1) return smartNotebook

        smartBookEntity.createdTimeMillis = updateTime
        smartBookEntity.lastModifiedTimeMillis = updateTime

        val bookId = smartBooksDao.insertSmartBook(smartBookEntity)
        smartBookEntity.bookId = bookId

        for (smartBookPage in smartNotebook.smartBookPages) {
            smartBookPage.bookId = bookId
            val id = smartBookPagesDao.insertSmartBookPage(smartBookPage)
            smartBookPage.id = id
        }
        // TODO: Shouldn't I save atomic notes as well?
        EventBus.getDefault().post(SmartNotebookSaved(smartNotebook, context))

        return smartNotebook
    }

    fun getSmartNotebookContainingNote(noteId: Long): MutableList<SmartNotebook?> {
        val pagesOfNote = smartBookPagesDao.getSmartBookPagesOfNote(noteId)
        if (pagesOfNote.isEmpty()) return ArrayList()

        return pagesOfNote.map { getSmartNotebooks(it.bookId) }
            .filter { it != null }
            .toMutableList()
    }

    val allSmartNotebooks: MutableList<SmartNotebook>
        get() {
            val smartBooks = smartBooksDao.allSmartBooks ?: return ArrayList()
            if (smartBooks.isEmpty()) return ArrayList()

            val smartNotebooks = ArrayList<SmartNotebook>()

            for (smartBook in smartBooks) {
                val smartBookPages = smartBookPagesDao.getSmartBookPages(smartBook.bookId)
                    ?: continue
                if (smartBookPages.isEmpty()) continue

                val noteIds = smartBookPages.map { it.noteId }.toMutableSet()

                val atomicNoteEntities =
                    (atomicNotesDomain.getAtomicNotes(noteIds) ?: continue).filterNotNull().toMutableList()
                smartNotebooks.add(SmartNotebook(smartBook, smartBookPages, atomicNoteEntities))
            }

            return smartNotebooks
        }

    fun getSmartNotebooksForNoteIds(noteIds: MutableSet<Long>): MutableSet<SmartNotebook> {
        val noteIdsLong = noteIds.toList().toMutableSet()
        val smartBookPages = smartBookPagesDao.getSmartBookPagesOfNote(noteIdsLong)
        val bookIdToPageMap =
            ListUtils.groupBy(smartBookPages) { it.bookId }
                .mapValues { it.value.toMutableList() }.toMutableMap()

        val bookIds = bookIdToPageMap.keys.toMutableSet()
        val smartBooks = smartBooksDao.getSmartBooks(bookIds)
        if (smartBooks.isEmpty()) return HashSet()

        val atomicNotes = atomicNotesDomain.getAtomicNotes(noteIdsLong).toList()
        val noteIdToNotes =
            ListUtils.groupBy(atomicNotes) { it.noteId }
                .mapValues { it.value.toMutableList() }.toMutableMap()

        val smartNotebooks = HashSet<SmartNotebook>()
        for (smartBook in smartBooks) {
            val bookPages = bookIdToPageMap[smartBook.bookId] ?: continue
            val bookNotes =
                bookPages.mapNotNull { page -> noteIdToNotes[page.noteId] }.flatten().toMutableList()
            smartNotebooks.add(SmartNotebook(smartBook, bookPages, bookNotes))
        }

        return smartNotebooks
    }

    fun getVirtualSmartNotebooks(bookTitle: String?, noteIds: MutableSet<Long>): SmartNotebook {
        val noteIdsLong = noteIds.toMutableSet()
        val atomicNoteEntities = atomicNotesDomain.getAtomicNotes(noteIdsLong)
        val smartBookPages = ArrayList<SmartBookPage>()

        var pageIndex = 0
        for (note in atomicNoteEntities) {
            smartBookPages.add(SmartBookPage(-1, note.noteId, pageIndex))
            pageIndex++
        }
        val smartBook = SmartBookEntity(
            -1, bookTitle,
            System.currentTimeMillis(),
            System.currentTimeMillis()
        )
        val notesList = atomicNoteEntities.toMutableList()
        return SmartNotebook(smartBook, smartBookPages, notesList)
    }

    fun getSmartNotebooks(bookId: Long): SmartNotebook? {
        val smartBook = smartBooksDao.getSmartbook(bookId) ?: return null
        val smartBookPages = smartBookPagesDao.getSmartBookPages(bookId)
        if (smartBookPages.isEmpty()) return null

        val noteIds = smartBookPages.map { it.noteId }.toMutableSet()

        val atomicNoteEntities = atomicNotesDomain.getAtomicNotes(noteIds)
        val smartNotebook = SmartNotebook(smartBook, smartBookPages, atomicNoteEntities)

        return smartNotebook
    }

    fun getSmartNotebooks(title: String): MutableSet<SmartNotebook> {
        if (title.length < 3) return HashSet()

        val smartBooks = smartBooksDao.getSmartbooksWithMatchingTitle("%$title%")
        if (smartBooks.isEmpty()) return HashSet()

        val bookIds = smartBooks.map { it.bookId }.toMutableSet()
        val smartBookPages = smartBookPagesDao.getSmartBooksPages(bookIds)
        val bookIdToPagesMap =
            ListUtils.groupBy(smartBookPages, { it.bookId })
                .mapValues { it.value.toMutableList() }.toMutableMap()

        val noteIds = smartBookPages.map { it.noteId }.toMutableSet()
        val atomicNotes = atomicNotesDomain.getAtomicNotes(noteIds)
        val noteIdToNotes =
            ListUtils.groupBy(atomicNotes,  { it.noteId })
                .mapValues { it.value.toMutableList() }.toMutableMap()

        val smartNotebooks = mutableSetOf<SmartNotebook>()
        for (smartBook in smartBooks) {
            val bookPages = bookIdToPagesMap[smartBook.bookId] ?: continue
            val bookNotes = bookPages.mapNotNull { page -> noteIdToNotes[page.noteId] }.flatten().toMutableList()
            smartNotebooks.add(SmartNotebook(smartBook, bookPages, bookNotes))
        }

        return smartNotebooks
    }

    private fun newSmartBook(title: String?, createdDateTimeMs: Long): SmartBookEntity {
        val smartBookEntity = SmartBookEntity()
        if (!isNullOrWhitespace(title)) {
            smartBookEntity.title = title
        }
        smartBookEntity.createdTimeMillis = createdDateTimeMs
        smartBookEntity.lastModifiedTimeMillis = createdDateTimeMs

        val bookId = smartBooksDao.insertSmartBook(smartBookEntity)
        smartBookEntity.bookId = bookId

        return smartBookEntity
    }

    fun newSmartBookPage(
        smartBookEntity: SmartBookEntity,
        atomicNoteEntity: AtomicNoteEntity,
        pageOrder: Int
    ): SmartBookPage {
        val smartBookPage = SmartBookPage(
            smartBookEntity.bookId,
            atomicNoteEntity.noteId,
            pageOrder
        )
        val id = smartBookPagesDao.insertSmartBookPage(smartBookPage)
        smartBookPage.id = id

        return smartBookPage
    }


    fun bookExists(notebook: SmartNotebook): Boolean {
        val smartBook = notebook.smartBook ?: return false
        val bookInDb = smartBooksDao.getSmartbook(smartBook.bookId)
        return bookInDb != null
    }

    val allSmartBookPages: MutableList<SmartBookPage>
        get() = smartBookPagesDao.allSmartBookPages
}



























